Web/Scotty.hs

```diff
@@ -113,7 +108,7 @@ setMaxRequestBodySize = Trans.setMaxRequestBodySize
 
 -- | Throw an exception, which can be caught with 'rescue'. Uncaught exceptions
 -- turn into HTTP 500 responses.
-raise :: Text -> ActionM a
+raise :: E.Exception e => e -> ActionM a
```

Comment:
Now that `ActionM` is an instance of `MonadUnliftIO`, users can use `throwIO` and `catch` from `unliftio`

Perhaps we could keep the type signatures of `raise` and `rescue` as-is for the sake of compatibility, and deprecate them away.

Reply:
This would mean specifying a catch-all exception type with a Text field. I briefly considered this design but it feels like a kludge, given that the new design supports custom exception types anyway. Also, what's the meaning of `catch`ing a string?

Reply:
Yes, I find it a strange design; that's why I suggest deprecating them. I think deprecation is more friendly to existing users than changing the type signature.

Reply:
Also, I don't think exporing the constructors of `ActionError` is a good idea, since currently there is no catching mechanism for those at the ScottyT level.

Reply:
Ah I see, it wraps the exception with `ActionError`. I now think this is a reasonable change.

Reply:
@fumieval I don't think we should deprecate `raise` either. Its behaviour is well-defined now, and I've restated its signature and original behaviour (throwing a 500 together with a string). However I've added a `throw = throwIO` function for convenience and adjusted tests and example programs.

Reply:
A consequence of this redesign, which I now realize the tests don't currently catch, is that "parameter not found" exceptions cannot be caught (and in fact nothing that is implemented in terms of `raiseStatus` can be caught either). Is this desirable?

Reply:
@fumieval  as of https://github.com/scotty-web/scotty/pull/314/commits/e9d852a3df1478ce24dcb6d92d4d3c4216129a4f there is a standalone `StatusError` which is meant to be `rescue`d (whereas Next, Redirect etc. cannot be caught).  But now this broke 4 tests. Would be great if you could take a look, otherwise I'll resume in the next days. Thanks!

Reply:
FTR the fix involved splitting a user-visible exception type `StatusError` (which can be thrown and caught by the user, in addition to library code) from `ActionError`.
This way, a user can recover from a missing query parameter but cannot recover from `next` or `finish` since they effectively asked for it.

I'd like to document the separation between user-defined and internal exceptions in the wiki, as well as the possibility for rescuing from certain exceptions in-line (inside an ActionT handler) vs globally.